### ISSUE TITLE
Missing working steps for Data Lake Storage Gen2 Containers

### ISSUE THREAD

**Issue by alainbryden** [MicrosoftDocs/sql-docs#5936]
Specifically [Section 4 Restore database to virtual machine from URL](https://docs.microsoft.com/en-us/sql/relational-databases/tutorial-use-azure-blob-storage-service-with-sql-server-2016?view=sql-server-2017#4----restore-database-to-virtual-machine-from-url)

```sql
RESTORE DATABASE SomeDatabase
    FROM URL = 'https://azeusanblbstranonimized.blob.core.windows.net/sql-storage/SomeDatabase.bak'
    WITH
        MOVE 'some_database_data' TO 'https://azeusanblbstranonimized.blob.core.windows.net/sql-storage/SomeDatabase.mdf',
        MOVE 'some_database_log'  TO 'https://azeusanblbstranonimized.blob.core.windows.net/sql-storage/SomeDatabase.ldf'
, REPLACE
```

These steps work perfectly using a Azure StorageV2 (general purpose v2) account with "Blob Service Containers",
![image](https://user-images.githubusercontent.com/2285037/105223652-6ff8c980-5b32-11eb-8bf7-23566e4f563e.png)
but if you opt to enable the "Data Lake Storage Gen2" feature (a.k.a. "Hierarchical namespace") when setting up your StorageV2 account, you get "Data Lake Storage" containers,
![image](https://user-images.githubusercontent.com/2285037/105223602-5e172680-5b32-11eb-9bb5-376a80d548ba.png)
which look and configure almost identically, but result in the following error:

> The operating system returned the error '13(The data is invalid.)' while attempting 'RestoreContainer::ValidateTargetForCreation' on 'https://azeusanblbstranonimized.blob.core.windows.net/sql-storage/SomeDatabase.mdf'. File 'some_database_data' cannot be restored to 'https://azeusanblbstranonimized.blob.core.windows.net/sql-storage/SomeDatabase.mdf'. Use WITH MOVE to identify a valid location for the file.

I don't know what the solution is, but this tutorial can no longer be followed with someone using the latest Azure Storage solution (DLS Gen2) :( 

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 15e8e46c-01f3-a76a-761f-5206e6d878e9
* Version Independent ID: 3945290e-ef38-b04c-9ee6-d3889b4b94e4
* Content: [Tutorial: Use Azure Blob storage service with SQL Server 2016 - SQL Server](https://docs.microsoft.com/en-us/sql/relational-databases/tutorial-use-azure-blob-storage-service-with-sql-server-2016?view=sql-server-2017)
* Content Source: [docs/relational-databases/tutorial-use-azure-blob-storage-service-with-sql-server-2016.md](https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/relational-databases/tutorial-use-azure-blob-storage-service-with-sql-server-2016.md)
* Product: **sql**
* Technology: ****
* GitHub Login: @MashaMSFT
* Microsoft Alias: **mathoma**

**Comment by WilliamAntonRohm** [comment]
@alainbryden -- Alain, thank you for your thorough feedback.  Please consider these resources:
- [SQL Server on Microsoft Q&A](https://docs.microsoft.com/en-us/answers/products/sql-server)
- [DBA Stack Exchange](https://dba.stackexchange.com/questions/tagged/sql-server)
- [Stack Overflow](https://stackoverflow.com/questions/tagged/sql-server) 

@MashaMSFT -- Masha, please look into this issue.

**Comment by alainbryden** [comment]
Thanks for triaging this - I do have a few days old stack-overflow question up (https://stackoverflow.com/questions/65813313) but it looks like I'll be the one posting the answer, if we ever get to the bottom of it.

I also opened issue #5938 for a similar tutorial that also suffers from this same error.

**Comment by MashaMSFT** [comment]
Hi @alainbryden , thanks for contacting Microsoft! This article covers [Azure blob storage](https://docs.microsoft.com/azure/storage/blobs/storage-blobs-overview). [Azure Data Lake](https://docs.microsoft.com/azure/storage/blobs/data-lake-storage-introduction) is a separate solution built on top of Azure Blob storage. SQL Server only supports page blob and block blobs, with block blobs being preferred. Azure Data Lake is not currently supported. 

I hope that helps and please let us know if we can be of further assistance. I'll close out this Git Issue for the time being but please feel free to comment should you need additional assistance. 

Thanks again for reaching out and I hope you have a wonderful rest of your day!
Masha from the SQL Docs team

**Comment by alainbryden** [comment]
I mean, ADL Containers specifically are supposed to be a complete
substitute for Blob. It supports both Page and Block blobs, so this
shouldnt be an issue.

They do maintain a list of "known issues" keeping it from reaching parity
here:
https://docs.microsoft.com/en-ca/azure/storage/blobs/data-lake-storage-known-issues
but this particular issue / limitation isn't listed. Perhaps it should be
added?

**Comment by bluefooted** [comment]
@alainbryden This is not a limitation of ADL, ADL is a full solution built on top of blob storage meant for big data analytics, it's not a relational database solution. SQL Server data and backup files have no way of taking advantage of the hierarchical file structure that ADL provides, so there would be no benefit of hosting SQL Server data files or backup files on ADL. This document gives an overview of what ADL is and how it works:
https://docs.microsoft.com/en-ca/azure/storage/blobs/data-lake-storage-introduction

Is there a specific feature of ADL that you were hoping to take advantage of that is missing in the existing SQL Server blob storage implementation?

**Comment by alainbryden** [comment]
From what I've seen "Azure Storage" is sold as a a single "unit" on the Azure Portal. See https://docs.microsoft.com/en-us/azure/storage/common/storage-introduction as a general example of client-facing materials.

> Azure Storage (Gen2) includes 5 "core" services (as a part of that **single** provisioned resource):
> - Azure Blobs
> - Azure Files
> - Azure Queues
> - Azure Tables
> - Azure Disks

Our project has licenced one Azure Storage instance for a couple year now and uses many of these features - we have a half a dozen blob containers, some tables, some queues, some file shares.

First our account was Gen1, and then it was "upgraded" to Gen2 as the latter got phased out - which went smoothly.

Now, Azure has expanded further on the capabilities of this resource and advertises AD DS (a.k.a. Hiererachical namespaces). The key value add is to authenticate with several services using Azure AD rather than SAS tokens. We opted to switch to purely to take advantage of those new features, with no expectation that any existing features would be "lost" - which is the general message one gets from the documentation and promotional materials.

Our use case for blob containers has not changed at all - and every aspect of the migration to the new Storage account went smoothly *except* for support for SQL files on blob, which mysteriously stopped working after the change. My guess is it's a bug, *not* an intentional lack of support as this seems to indicate, and that support will eventually be fixed. The first step is to create recognition of that oversight though - and that's proving to be quite difficult.

**Comment by alainbryden** [comment]
Got a reply from paid microsoft support on this.

> 
> Hello Alain,
> 
> We were able to complete the investigation and confirm we cannot use ADLS to store SQL Server datafiles because it uses Block Blob. We only use Page Blobs for storing data on Azure Storage due to requirement for Random writes.
> 
> We are working on updating the document to explicitly document the same about the supportability of ADLS Gen2.
> 
> Thank you once again for bring this to us and we hope to have this feature in future.

tl;dr Gen2's Blob storage has as a [known issue](https://docs.microsoft.com/en-ca/azure/storage/blobs/data-lake-storage-known-issues) that it does not yet support the Blob REST APIs for `PUT Page` and `GET Page Ranges`. SQL uses these APIs behind the scenes, hence the feature breaks when pointed towards Gen2 containers.

**Comment by alainbryden** [comment]
> SQL Server only supports page blob and block blobs, with block blobs being preferred.

@MashaMSFT Unfortunately, it turns out your above statement is not true, and therein lies the nature of the bug.

ADLS Gen2 has full support for block blobs, but SQL (whether documentation realizes it or not) **requires** use of Page blobs in order to support this feature. Whoever documented that "block blobs are preferred" must not be aware of this limitation, or that behind-the-scenes, Page blobs are still being used regardless.

**Comment by alainbryden** [comment]
There is only one other piece of documentation I've found on MicrosoftDocs that alludes to the feature of storing SQL files on blob: [SQL Server data files in Microsoft Azure](https://docs.microsoft.com/en-us/sql/relational-databases/databases/sql-server-data-files-in-microsoft-azure)

In that case, there's actually a hint in the documentation of where things might go south:
> This new feature uses Page blobs, which are more efficient when ranges of bytes in a file are modified frequently.

Perhaps putting a similar disclosure near at the top of this documentation (i.e. in the "Prerequisites" section) would avoid someone going down this path with an incompatible blob storage account.

In both cases (this article and the other article), a clear statement that the blob storage containers offered by ADLS Gen2 storage are **not** compatible due to their lack of support for the "Page Blob" APIs would be extraordinarily helpful.

**Comment by bluefooted** [comment]
@alainbryden I have updated that document that you just referenced to indicate that we do not support ADLS, it should go live later this morning. Also as you noted in one of your comments above, ADLS does not benefit SQL Server workloads, so there is no intention on our part to support storing SQL Server data files on ADLS.

**Comment by bluefooted** [comment]
@alainbryden Also to be clear, this page that you linked to did already specify that we use page blobs for this type of storage, the comment about block blobs being preferred is only in reference to backup to URL. With backups we are streaming to the blob storage and appending to the target file, an I/O pattern that generally performs better against block blobs. With data files, we are randomly reading and writing pages so page blobs perform better.

**Comment by alainbryden** [comment]
Thanks for the clarification on the other page @bluefooted 

I humbly suggest that it might also be worth adding a small disclaimer in [Section 4](https://docs.microsoft.com/en-us/sql/relational-databases/tutorial-use-azure-blob-storage-service-with-sql-server-2016?view=sql-server-ver15#4----restore-database-to-virtual-machine-from-url) of the current page, right before the sample code:

```sql
 MOVE 'AdventureWorks2016_data' to 'https://<mystorageaccountname>.blob.core.windows.net/<mystorageaccountcontainername>/AdventureWorks2016_Data.mdf'  
,MOVE 'AdventureWorks2016_log' to 'https://<mystorageaccountname>.blob.core.windows.net/<mystorageaccountcontainername>/AdventureWorks2016_Log.ldf'  
```

Which (for the first time in this guide) would require use of the Page Blobs APIs to `MOVE` MDF/LDF files onto blob.

I agree that it's probably not worth adding a disclaimer in the "Prerequisites" section at the top of this one, because this article is predominantly regarding backups (which continue to work on ADLS Gen2) - if not for this small section describing the process of restoring the database with backing files still on blob.

**Comment by bluefooted** [comment]
@alainbryden In this tutorial you're using the same storage account for both the backups and the data files, so you don't really want to turn on hierarchical files on the account. We don't have detailed steps to create the account and the link that's there does include a note about ADLS. I can add this information to the pre-requisites bullet, but by the time you get to section 4 it's too late to point this out :)
